pkg/workflow/transitioner.go

```diff
@@ -0,0 +1,262 @@
+			}
+			if serr.Status().Code == http.StatusConflict {
+				return true, requeueAfterStatusConflictTime, fmt.Errorf("encountered status conflict on workflow update (%v), requeuing after %v", workflow.Name, requeueAfterStatusConflictTime)
+			}
+			// TODO We probably want this to be larger than 0.
+			return true, 0, serr
+		}
+	}
+	return false, 0, nil
+}
+
+// TODO: We might want to move this to another file. Maybe utils?
+func (t *Transitioner) setLabels(workflow *api.Workflow) (newWorkflow *api.Workflow, err error) {
+	glog.V(3).Infof("Setting labels on wf %v", workflow.Name)
+	if workflow.Labels == nil {
+		workflow.Labels = make(map[string]string)
+	}
+	if workflow.Spec.JobsSelector == nil {
+		workflow.Spec.JobsSelector = &k8sApiUnv.LabelSelector{
+			MatchLabels: make(map[string]string),
+		}
+	}
+	workflow.Labels[workflowUID] = string(workflow.UID)
+	workflow.Spec.JobsSelector.MatchLabels[workflowUID] = string(workflow.UID)
+	newWorkflow, err = t.tpClient.Workflows(workflow.Namespace).Update(workflow)
+	return
+}
+
+// pastActiveDeadline checks if workflow has ActiveDeadlineSeconds field set and if it is exceeded.
+func pastActiveDeadline(workflow *api.Workflow) bool {
+	return false
+}
+
+func (t *Transitioner) updateWorkflow(workflow *api.Workflow) error {
+	_, err := t.tpClient.Workflows(workflow.Namespace).Update(workflow)
+	return err
+}
+
+func (t *Transitioner) manageWorkflow(workflow *api.Workflow) bool {
+	needsStatusUpdate := false
+	glog.V(3).Infof("Manage workflow %v", workflow.Name)
+	workflowComplete := true
+	for stepName, step := range workflow.Spec.Steps {
+		if stepStatus, ok := workflow.Status.Statuses[stepName]; ok && stepStatus.Complete {
+			continue // step completed nothing to do
+		}
+		workflowComplete = false
+		switch {
+		case step.JobTemplate != nil: // Job step
+			needsStatusUpdate = t.manageWorkflowJob(workflow, stepName, &step) || needsStatusUpdate
+		case step.ExternalRef != nil: // external object reference
+			needsStatusUpdate = t.manageWorkflowReference(workflow, stepName, &step) || needsStatusUpdate
+		}
+	}
+
+	if workflowComplete {
+		glog.V(3).Infof("Setting workflow complete status for workflow %v", workflow.Name)
+		now := k8sApiUnv.Now()
+		condition := api.WorkflowCondition{
+			Type:               api.WorkflowComplete,
+			Status:             k8sApi.ConditionTrue,
+			LastProbeTime:      now,
+			LastTransitionTime: now,
+		}
+		workflow.Status.Conditions = append(workflow.Status.Conditions, condition)
+		workflow.Status.CompletionTime = &now
+		needsStatusUpdate = true
+	}
+
+	return needsStatusUpdate
+}
+
+// TODO: I think this method should return an error (instead of logging the error).
+//       hen an error is returned we probably want to requeue the workflow.
+// manageWorkflowJob manages a Job that is in a step of a workflot.
+func (t *Transitioner) manageWorkflowJob(workflow *api.Workflow, stepName string, step *api.WorkflowStep) bool {
```

Comment:
I'm not sure if manage is the right term here. It seems that by calling it a transitioner we might want to stick with that terminlology? Maybe handleWorkflowJob or processWorkflowJob?